tt_metal/graph_tracking.cpp

```diff
@@ -0,0 +1,90 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "tt_metal/graph_tracking.hpp"
+
+
+using namespace tt::tt_metal;
+
+
+size_t GraphTracker::add_processor(const std::shared_ptr<IGraphProcessor>& new_processor) {
+    processors.push_back(new_processor);
+    return processors.size();
+}
+
+bool GraphTracker::add_hook(const std::shared_ptr<IGraphHooks>& new_hook) {
+    if (hook) {
+        return false;
+    }
+    hook = new_hook;
+    return true;
+}
+
+void GraphTracker::track_allocate(Buffer* buffer, bool bottom_up) {
+    if (processors.empty()) {
+        return;
+    }
+    for (auto& it : processors) {
+        it->track_allocate(buffer, bottom_up);
+    }
+}
+void GraphTracker::track_deallocate(Buffer* buffer) {
+    if (processors.empty()) {
+        return;
+    }
+    for (auto& it : processors) {
+        it->track_deallocate(buffer);
+    }
+}
+
+void GraphTracker::track_allocate_cb(const CoreRangeSet &core_range_set, uint64_t addr, uint64_t size) {
+    if (processors.empty()) {
+        return;
+    }
+    for (auto& it : processors) {
+        it->track_allocate_cb(core_range_set, addr, size);
+    }
+}
+
+void GraphTracker::track_deallocate_cb() {
+    if (processors.empty()) {
+        return;
+    }
+    for (auto& it : processors) {
+        it->track_deallocate_cb();
+    }
+}
+
+bool GraphTracker::hook_allocate(Buffer* buffer, bool bottom_up) {
+    if (hook == nullptr)
+        return false;
+
+    return hook->hook_allocate(buffer, bottom_up);
+}
+
+bool GraphTracker::hook_deallocate(Buffer* buffer) {
+    if (hook == nullptr)
+        return false;
+    return hook->hook_deallocate(buffer);
+}
+
+bool GraphTracker::block_run_program() {
```

Comment:
`block_run_program()` looks like an action. "Block it!"

Consider 
* bool is_program_run_blocked() const;
* bool should_block_program_run() const;
or even
* hook_program_run(.. ) which can completely override how the program is run

Looks like should be `const`

Reply:
desided to rename to hook for now.